text/0008-ingest-gem-content.md

```diff
@@ -0,0 +1,332 @@
+Binary files will be recorded in metadata, but not stored.
+Gem contents will be stored at an S3 key named for the SHA256 of the file to de-duplicate files within a gem.
+Yanking will remove file contents unique to the gem, as expected, without altering unyanked gem contents.
+
+### Background process
+
+In order to ingest the contents of a gem, a background process can be enqueued to index and store the gem contents.
+
+At gem upload time, or on demand, a background worker will download, unpack, record file metadata, index files in the database as rows, and upload each file to a content-only S3 bucket.
+The process will result in a row in the database for each file in a gem, with the relative path, checksum, and size of every file.
+
+### Indexing files to create a manifest
+
+The background worker will create a json document manifest of all the files and store it in S3 alongside the files themselves.
+
+The manifest will hold the path, checksum, size, content type, binary determination, and plain text line count.
+
+In order to support yanking gems, which requires comparing checksums across all versions of a gem, a table will be added to the rubygems.org database.
+The table will consist of one row per gem version with a foreign key for the `version_id` and a jsonb column holding only the SHA256 checksums of all files in that version.
+
+The table will be indexed on `version_id` to support looking up all checksums in all versions of a gem.
+This table is for the sole purpose of yanking a gem, discussed below.
+
+### Recording, but not storing binary files
+
+Files that are identified as binary will be flagged in the index and not uploaded to the storage bucket.
+This is both to save storage space and avoid potential abuse.
+Large compiled binaries, image data, or other binary data is not easily browsed or diffed.
+Both npmjs.com's code browsing and github.com excludes binary files from display.
+
+### Using SHA256 to reduce storage size and increase cache hits
+
+As of this writing, there is more than 5TB of gem files stored by rubygems.org.
+
+Using `rake` as an example, the total size of all `.gem` files is currently 6.8MB.
+
+The average compression across the entire history of the rake gem is about 84%.
+Expanded fully, all versions of rake add up to 42MB.
+
+Many files in rake do not change between versions, allowing us to optimize the storage.
+If each file is instead saved at its SHA256 checksum, causing and identical file to be stored only once, the size of all uncompressed files in `rake` droms from 42MB to 10MB.
+The resulting directory has 1226 files for all versions of `rake`.
+
+In order to access a hashed filename, access the file content table to find all files in the version of rake.
+Use the relative file path within the gem to find the SHA256 checksum.
+Access the S3 bucket at a key similar to `/rake/files/c588eda0547f6fa72fd7cd9ba2ddc2e81b96e61d690a929e8f579412da27eebd`
```

Comment:
should the key include the gem name? we might achieve significantly better de-duplication if we only key on the content and nothing else, since I'd imagine there are some files duplicated across gems

Reply:
I discussed this in the rationale section. Briefly: the deduplication benefit requires all files in all gems be searched for duplicates when deleting any gem. This requires one row per file, which was my original proposal. The _only_ benefit that I can see in one-row-per-file is this minor deduplication. I expect it to be a tiny benefit since it will only deduplicate gems vendored in other gems (I expect this to be a very small portion of total gems) and files that are frequently the same, like LICENSE.

I also think it significantly increases the yank/push race condition situation since now any gem, across all gems, could trigger the race instead of just versions within the same gem (which I still don't think is insignificant, but at least it's a small enough number of files that we can verify all of a gem's content files after it's yanked).

Reply:
Rather than reply to another related comment below about one row per file, I will attach here my response to your comment below:

I have a part of the document I haven't committed yet that tries to deal with this particular race condition. I'll upload it soon (it's just a little messy, working on that today). It's relevant whether we group by gems or globally. I'm not sure locking individual file rows would be sufficient, though I haven't been able to completely solve the race condition yet without just "re-checking" s3 at the end of a process and re-entering the entire process if anything is incorrect. I'd be happy to hear your thoughts on how to approach this.

I'm also curious if you have a different view of the benefit of the additional deduplication. If I'm wrong about how much space it would save, it could be worth the additional database load and size. I'd hate to simply trade saving a few 100MB on S3 for adding few 100MB index to the postgres database. The only purpose of the index is for yanking gems without a 100MM row table scan, and yet the index could arguably be larger than the savings on S3. Please do correct my if you know of a more significant deduplication savings than I'm estimating that could counteract that. 

Originally I had planned to store a row per file for every file that included all of the information, like content-type and size. That does not optimize for loading the manifest which is vastly more common than yanking gems. We rewrote the proposal for storing the manifest statically on S3 and only storing the hashes necessary for yanks in the database. It's even arguable that yank is uncommon enough that we could do the entire process by loading all the static manifests from S3 for every gem (they'll all be in a single directory within the gem). Then any locking or race condition solution would be entirely managed through S3. I haven't thought it through yet, but I don't see why that wouldn't be possible.